nat/behavior/service.go

```diff
@@ -31,18 +34,32 @@ var compatibleSTUNServers = []string{
 
 const concurrentRequestTimeout = 1 * time.Second
 
+// ErrInappropriateState error is returned by gatedNATProber when connection
+// is active
+var ErrInappropriateState = errors.New("NAT probing is impossible at this connection state")
+
 // NATProber is an abstaction over instances capable probing NAT and
 // returning either it's type or error.
 type NATProber interface {
 	Probe(context.Context) (string, error)
 }
 
+// ConnectionStatusProvider is a subset of connection.Manager methods
+// to provide gatedNATProber with current connection status
+type ConnectionStatusProvider interface {
+	Status() connectionstate.Status
+}
+
 // NewNATProber constructs some suitable NATProber without any implementation
 // guarantees.
-func NewNATProber() NATProber {
-	return newConcurrentNATProber(compatibleSTUNServers, concurrentRequestTimeout)
+func NewNATProber(connStatusProvider ConnectionStatusProvider) NATProber {
+	var prober NATProber
+	prober = newConcurrentNATProber(compatibleSTUNServers, concurrentRequestTimeout)
+	prober = newGatedNATProber(connStatusProvider, prober)
+	return prober
```

Comment:
Can we make it like this? I was confusing me for some time all these assignment to the same var. 
```suggestion
	prober := newConcurrentNATProber(compatibleSTUNServers, concurrentRequestTimeout)
	return newGatedNATProber(connStatusProvider, prober)
```

Reply:
We can, but we have no need to. When we'll need to add yet another "prober" into this chain (e.g. one which hooks on result and sends metrics) it will be more readable if we'll have it like this:

```
+    prober = newSnoopingNATProber(prober, qualitySender)
```